java/ql/src/experimental/Security/CWE/CWE-094/InsecureBeanValidation.ql

```diff
@@ -0,0 +1,76 @@
+/**
+ * @name Insecure Bean validation
+ * @description User-controlled data may be evaluated as a Java EL expressions, leading to arbitrary code execution.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id java/unsafe-eval
+ * @tags security
+ *       external/cwe/cwe-094
+ */
+
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.FlowSources
+import DataFlow::PathGraph
+
+class BeanValidationSource extends RemoteFlowSource {
+  BeanValidationSource() {
+    exists(Method m, Parameter v |
+      this.asParameter() = v and
+      m.getParameter(0) = v and
+      m
+          .getDeclaringType()
+          .getASupertype+()
+          .getSourceDeclaration()
+          .hasQualifiedName("javax.validation", "ConstraintValidator") and
+      m.hasName("isValid") and
+      m.fromSource()
+    )
+  }
+
+  override string getSourceType() { result = "BeanValidation source" }
+}
+
+class ExceptionTaintStep extends TaintTracking::AdditionalTaintStep {
+  override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
+    exists(Call call, TryStmt t, CatchClause c, MethodAccess gm |
+      call.getEnclosingStmt().getEnclosingStmt*() = t.getBlock() and
+      t.getACatchClause() = c and
+      (
+        call.getCallee().getAThrownExceptionType().getASubtype*() = c.getACaughtType() or
+        c.getACaughtType().getASupertype*() instanceof TypeRuntimeException
+      ) and
+      c.getVariable().getAnAccess() = gm.getQualifier() and
+      gm.getMethod().getName().regexpMatch("get(Localized)?Message|toString") and
+      n1.asExpr() = call.getAnArgument() and
+      n2.asExpr() = gm
+    )
+  }
+}
```

Comment:
This taint step seems to ad-hoc, please delete.

Reply:
ok, will do, but that taintstep was required for the Netflix findings. In some cases, the way to validate a bean property is parsing it and consider it valid if it does not through an exception. Otherwise include the exception message in the bean validation error message. In that case, if the user input is reflected in the exception message, then it will lead to a real issue.

Reply:
Taint flow through exceptions definitely is one of our blind spots, so I can understand the need for something like this. I would just like it to be a little more rigorous and then applied to all queries.